sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala

```diff
@@ -0,0 +1,222 @@
+      digitChar: Expression) =
+    this(input, upperChar, lowerChar, digitChar, Literal(Mask.MASKED_IGNORE))
+
+  /**
+   * Expected input types from child expressions. The i-th position in the returned seq indicates
+   * the type requirement for the i-th child.
+   *
+   * The possible values at each position are:
+   *   1. a specific data type, e.g. LongType, StringType. 2. a non-leaf abstract data type, e.g.
+   *      NumericType, IntegralType, FractionalType.
+   */
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(StringType, StringType, StringType, StringType, StringType)
+
+  /**
+   * Called by default [[eval]] implementation. If subclass of QuinaryExpression keep the default
+   * nullability, they can override this method to save null-check code. If we need full control
+   * of evaluation process, we should override [[eval]].
+   */
+  override protected def nullSafeEval(
+      input: Any,
+      upperChar: Any,
+      lowerChar: Any,
+      digitChar: Any,
+      otherChar: Any): Any =
+    Mask.transformInput(
+      input.asInstanceOf[UTF8String],
+      upperChar.asInstanceOf[UTF8String],
+      lowerChar.asInstanceOf[UTF8String],
+      digitChar.asInstanceOf[UTF8String],
+      otherChar.asInstanceOf[UTF8String])
+
+  /**
+   * Returns Java source code that can be compiled to evaluate this expression. The default
+   * behavior is to call the eval method of the expression. Concrete expression implementations
+   * should override this to do actual code generation.
+   *
+   * @param ctx
+   *   a [[CodegenContext]]
+   * @param ev
+   *   an [[ExprCode]] with unique terms.
+   * @return
+   *   an [[ExprCode]] containing the Java source code to generate the given expression
+   */
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode =
+    defineCodeGen(
+      ctx,
+      ev,
+      (input, upperChar, lowerChar, digitChar, otherChar) => {
+        s"org.apache.spark.sql.catalyst.expressions.Mask." +
+          s"transformInput($input, $upperChar, $lowerChar, $digitChar, $otherChar);"
+      })
+
+  /**
+   * Returns the [[DataType]] of the result of evaluating this expression. It is invalid to query
+   * the dataType of an unresolved expression (i.e., when `resolved` == false).
+   */
+  override def dataType: DataType = StringType
+
+  /**
+   * Returns a Seq of the children of this node. Children should not change. Immutability required
+   * for containsChild optimization
+   */
+  override def children: Seq[Expression] =
+    Seq(input, upperChar, lowerChar, digitChar, otherChar)
+
+  override protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Mask =
+    copy(
+      input = newChildren(0),
+      upperChar = newChildren(1),
+      lowerChar = newChildren(2),
+      digitChar = newChildren(3),
+      otherChar = newChildren(4))
+}
+
+case class MaskArgument(maskChar: Char, ignore: Boolean)
+
+object Mask {
+  private val MASKED_UPPERCASE = 'X'
```

Comment:
please add a comment for each of these private vals mentioning what the character represents?